// office-addin-sso/lib/msgraph-helper.js
// Copyright (c) Microsoft. All rights reserved. Licensed under the MIT license. See full license in the root of the repo.
/*
    This file provides the provides functionality to get Microsoft Graph data.
*/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeGraphApiCall = exports.getGraphToken = exports.getGraphData = void 0;
const odata_helper_1 = require("./odata-helper");
const message_helper_1 = require("./message-helper");
const $ = require("jquery");
const domain = 'graph.microsoft.com';
const versionURLsegment = '/v1.0';
function getGraphData(accessToken, apiURLsegment, queryParamsSegment) {
    return __awaiter(this, void 0, void 0, function* () {
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            try {
                const oData = yield odata_helper_1.ODataHelper.getData(accessToken, domain, apiURLsegment, versionURLsegment, queryParamsSegment);
                resolve(oData);
            }
            catch (err) {
                reject(`Error get Graph data. \n${err}`);
            }
        }));
    });
}
exports.getGraphData = getGraphData;
function getGraphToken(bootstrapToken) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            let response = yield $.ajax({
                type: "GET",
                url: "/auth",
                headers: { Authorization: "Bearer " + bootstrapToken },
                cache: false
            });
            return response;
        }
        catch (err) {
            throw new Error(`Error getting Graph token. \n${err}`);
        }
    });
}
exports.getGraphToken = getGraphToken;
function makeGraphApiCall(accessToken) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const response = yield $.ajax({
                type: "GET",
                url: "/getuserdata",
                headers: { access_token: accessToken },
                cache: false
            });
            return response;
        }
        catch (err) {
            message_helper_1.showMessage(`Error from Microsoft Graph. \n${err}`);
        }
    });
}
exports.makeGraphApiCall = makeGraphApiCall;
